File: backend/src/scheduler/core/reward.ts
import { MOVE_REWARD_SCALE_MINUTES } from "../constants";

/**
 * Pure delayed-reward math for the first user `MOVE` of a LinUCB-placed
 * session (ADR-0001 §7/§9). No I/O, no clock.
 */

/**
 * Grades a drag/resize/pick displacement into a LinUCB reward in `[-1, 0]`:
 * `0` for a zero-distance move (e.g. a resize that doesn't shift the start),
 * otherwise a penalty that grows linearly with `|dragDistanceMinutes|` and
 * saturates at `-1` once the displacement reaches `MOVE_REWARD_SCALE_MINUTES`.
 */
export function dragDistanceReward(
  dragDistanceMinutes: number,
  scaleMinutes: number = MOVE_REWARD_SCALE_MINUTES,
): number {
  if (dragDistanceMinutes === 0) return 0;
  return -Math.min(1, Math.abs(dragDistanceMinutes) / scaleMinutes);
}

File: services/bandit/src/core/reward.py
"""Graded MOVE reward (port of ``reward.ts``)."""

from .constants import MOVE_REWARD_SCALE_MINUTES


def drag_distance_reward(
    drag_distance_minutes: float,
    scale_minutes: float = MOVE_REWARD_SCALE_MINUTES,
) -> float:
    if drag_distance_minutes == 0:
        return 0.0
    return -min(1.0, abs(drag_distance_minutes) / scale_minutes)
